src/couchdb/rolling-database.ts

```diff
@@ -0,0 +1,141 @@
+import {
+  DocumentInsertResponse,
+  MangoSelector,
+  ServerScope,
+  SortOrder
+} from 'nano'
+
+import { setupDatabase, SetupDatabaseOptions } from '..'
+import { CouchDoc } from './as-couch-doc'
+import { RollingDatabasesDocument } from './rolling-database-document'
+import { DatabaseSetup } from './setup-database'
+import { SyncedDocument } from './synced-document'
+
+/**
+ * Describes a rolling collection of Couch databases that should exist.
+ */
+export type RollingDatabaseSetup = Pick<
+  DatabaseSetup,
+  'options' | 'replicatorSetup' | 'documents' | 'templates'
+>
+
+/**
+ * Arguments to the rolling database query.
+ */
+export interface RollingFindOptions {
+  // How many documents to return. Defaults to 20:
+  limit?: number
+
+  // How far in the past we should look. Defaults to no limit:
+  startDate?: Date
+
+  // The Mango sort order:
+  sort?: SortOrder[]
+}
+
+export interface RollingDatabase {
+  // Executes a query against the stack of databases:
+  find: (
+    selector: MangoSelector,
+    opts?: RollingFindOptions
+  ) => Promise<Array<CouchDoc<object> & { db: string }>>
+
+  // Inserts a document into the newest non-future database:
+  insert: (
+    doc: CouchDoc<object> & { db?: string }
+  ) => Promise<DocumentInsertResponse>
+
+  setup: (
+    setupInfo: RollingDatabaseSetup,
+    opts?: SetupDatabaseOptions
+  ) => Promise<void>
+}
+
+export async function makeRollingDatabase(
+  connection: ServerScope,
+  databaseList: SyncedDocument<RollingDatabasesDocument>
+): Promise<RollingDatabase> {
+  // We only want to set up each database once:
+  const ready: { [name: string]: true } = {}
+
+  return {
+    async find(selector, opts = {}) {
+      const { limit = 20, startDate, sort } = opts
+
+      const out: Array<CouchDoc<object> & { db: string }> = []
+      for (const { name, start } of databaseList.doc.databases) {
+        if (!ready[name]) continue
+
+        const { docs } = await connection.use(name).find({
+          limit: limit - out.length,
+          selector,
+          sort
+        })
+        for (const doc of docs) {
+          const { _id, _rev, ...rest } = doc
+          out.push({ id: _id, rev: _rev, db: name, doc: rest })
```

Comment:
Seems like we should be using the `asCouchDoc` cleaner here?

Reply:
Same problem, because it's just `asUnknown`.